File: 05-Algorithms/code/python/seb_alg/dijkstra.py
"""Mirrors `typescript/src/dijkstra.ts` exactly."""

from __future__ import annotations

import math
from typing import List, Optional, Sequence

from .errors import AlgoError
from .graph_common import build_weighted_adj_list, check_vertex_in_range


def dijkstra(n: int, edges: Sequence[Sequence[int]], src: int, directed: bool = False) -> List[Optional[int]]:
    """Single-source shortest paths, O(V^2 + E) array-scan Dijkstra.
    Requires non-negative weights. Unreachable vertices are None.
    """
    check_vertex_in_range(n, src)
    for _, _, w in edges:
        if w < 0:
            raise AlgoError("invalid", "dijkstra requires non-negative edge weights")
    adj = build_weighted_adj_list(n, edges, directed)
    dist: List[float] = [math.inf] * n
    dist[src] = 0
    visited = [False] * n

    for _ in range(n):
        u = -1
        for v in range(n):
            if not visited[v] and dist[v] < math.inf and (u == -1 or dist[v] < dist[u]):
                u = v
        if u == -1:
            break
        visited[u] = True
        for v, w in adj[u]:
            if dist[u] + w < dist[v]:
                dist[v] = dist[u] + w

    return [None if d == math.inf else d for d in dist]

File: 05-Algorithms/code/typescript/src/dijkstra.ts
import { AlgoError } from "./errors.js";
import { buildWeightedAdjList, checkVertexInRange, type WeightedEdge } from "./graphCommon.js";

/**
 * Single-source shortest paths with a binary-heap-free O(V^2 + E) Dijkstra
 * (a plain array scan for the minimum each round; fine for the small graphs
 * in this lab and avoids a second data structure entirely). Requires all
 * weights to be non-negative (throws AlgoError("invalid") otherwise).
 * Unreachable vertices are `null` in the result (rather than +Infinity), and
 * `dist[src] === 0`.
 */
export function dijkstra(n: number, edges: WeightedEdge[], src: number, directed = false): (number | null)[] {
  checkVertexInRange(n, src);
  for (const [, , w] of edges) {
    if (w < 0) throw new AlgoError("invalid", "dijkstra requires non-negative edge weights");
  }
  const adj = buildWeightedAdjList(n, edges, directed);
  const dist = new Array(n).fill(Infinity);
  dist[src] = 0;
  const visited = new Array(n).fill(false);

  for (let iter = 0; iter < n; iter++) {
    let u = -1;
    for (let v = 0; v < n; v++) {
      if (!visited[v] && dist[v] < Infinity && (u === -1 || dist[v] < dist[u])) u = v;
    }
    if (u === -1) break;
    visited[u] = true;
    for (const [v, w] of adj[u]) {
      if (dist[u] + w < dist[v]) dist[v] = dist[u] + w;
    }
  }

  return dist.map((d) => (d === Infinity ? null : d));
}
